File: src/game/controller/MouseControllerManager.java
package game.controller;

import game.Bomberman;
import game.engine.PeriodicTask;
import game.entity.Player;
import game.entity.models.Entity;
import game.models.Coordinates;
import game.models.Direction;

import java.awt.event.*;
import java.util.ArrayList;
import java.util.LinkedList;

public class MouseControllerManager extends MouseAdapter implements MouseMotionListener {
    public boolean mouseDraggedInteractionOccured;
    public boolean mouseDraggedInteractionInterrupted;
    public boolean mouseClickedInteractionSuccessful;
    public boolean mouseDragInteractionEntered = false;
    public boolean isMouseDragged;
    public boolean isMouseClicked;
    private LinkedList<MouseEvent> mouseCLickQueue = new LinkedList<>();
    //first and last directions from player are differentiated so the task can be terminated when they have no direction in common, in order to avoid infinite loop
    private ArrayList<Direction> firstDirectionsFromPlayer = new ArrayList<>();
    ArrayList<Direction> latestDirectionsFromPlayer =new ArrayList<>();
    public Coordinates mouseCoords;
    private final int DELAY = 300;
    private Entity entity = null;

    private final Runnable task = () -> {
        Player currPlayer = Bomberman.getMatch().getPlayer();
        if(currPlayer==null)return;
        onCooldown();
        //interact with mouse
        if (Bomberman.getMatch().getPlayer() != null && !currPlayer.getAliveState())
            return;


        if (entity != null && Bomberman.getMatch().getPlayer().getListClassInteractWithMouse().contains(entity.getClass())) {
            entity.mouseInteractions();
            onPeriodicTaskEnd();
            nextCommandInQueue();
            return;
        }


        //TODO unregister observer
        //else move with mouse
        else {
            latestDirectionsFromPlayer = mouseCoords.fromCoordinatesToDirection(currPlayer.getCoords());
            for (Direction d : latestDirectionsFromPlayer) {
                Bomberman.getMatch().getControllerManager().onKeyPressed(d.toCommand());
            }

            if (firstDirectionsFromPlayer.isEmpty()
                    || firstDirectionsFromPlayer.stream().anyMatch(d -> latestDirectionsFromPlayer.contains(d))
            ) {
                return;
            }
            onPeriodicTaskEnd();
            nextCommandInQueue();
        }

    };
    public void nextCommandInQueue(){
        mouseCLickQueue.poll();
        if(!mouseCLickQueue.isEmpty())
        onMouseClicked(mouseCLickQueue.peekFirst());
    }
    @Override
    public void mouseReleased(MouseEvent event){

        //if mouse dragged interaction was not successful and mouse is released, call mouseClicked so that click-dependent interactions can be called.
        //this is necessary because mouseClicked is never called when mouseDragged is called, unlike mousePressed. MouseClicked was still chosen because
        // the two methods shouldn't be called at exactly the same time.
        if(isMouseDragged) {
            if (!mouseDraggedInteractionOccured) mouseClicked(event);
            mouseDragInteractionEntered = false;
            isMouseDragged = false;
            mouseDraggedInteractionOccured = false;
            mouseDraggedInteractionInterrupted = false;
        }



    }
    public void onMouseReleased(){

    }
    @Override
    public void mousePressed(MouseEvent event){
        if(isMouseClicked) return;
        mouseCoords = new Coordinates(event.getX(), event.getY());
        entity = Coordinates.getEntityOnCoordinates(mouseCoords);
    }


    public PeriodicTask playerMovementTask = new PeriodicTask(task,DELAY);

    @Override
    public void mouseClicked(MouseEvent e) {
        if(!mouseCLickQueue.isEmpty()) {
            mouseCLickQueue.add(e);
            return;

        }
        mouseCLickQueue.add(e);
        onMouseClicked(e);

    }
    public void onMouseClicked(MouseEvent e){
        if(isMouseClicked) return;
        isMouseClicked=true;

        if(Bomberman.getMatch().getPlayer()==null|| !Bomberman.getMatch().getPlayer().getAliveState()) {
            return;
        }
        //isMouseClicked = true;
        mouseCoords = new Coordinates(e.getX(), e.getY());
        firstDirectionsFromPlayer = mouseCoords.fromCoordinatesToDirection(Bomberman.getMatch().getPlayer().getCoords());
        entity = Coordinates.getEntityOnCoordinates(mouseCoords);
        playerMovementTask.resume();
    }


    //directions are refreshed and will be replaced in task
    public void onCooldown(){
        for (Direction d: firstDirectionsFromPlayer) {
            Bomberman.getMatch().getControllerManager().onKeyReleased(d.toCommand());
        }
    }
    public void stopPeriodicTask(){
        mouseCLickQueue = new LinkedList<>();
        onPeriodicTaskEnd();
    }
    public void onPeriodicTaskEnd(){

        if(!isMouseClicked) return;
        playerMovementTask.stop();
        onCooldown();
        isMouseClicked = false;

    }
    @Override
    public void mouseDragged(MouseEvent event){
        if(isMouseClicked) return;

        isMouseDragged = true;
        mouseCoords = new Coordinates(event.getX(), event.getY());
        if (entity==null)return;
        entity.mouseInteractions();

    }

    public Entity getEntity(){
        return entity;
    }

}

File: src/game/input/MouseControllerManager.kt
package game.input

import game.JBomb
import game.JBomb.match
import game.domain.tasks.PeriodicTask
import game.domain.world.domain.entity.actors.abstracts.base.Entity
import game.domain.world.domain.entity.actors.impl.bomber_entity.player.Player
import game.domain.world.domain.entity.geo.Coordinates
import game.domain.world.domain.entity.geo.Direction
import kotlinx.coroutines.CoroutineScope
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionListener
import java.util.*

/**
 * The MouseControllerManager class handles mouse interactions and controls player movement based on mouse input.
 */
class MouseControllerManager(
        scope: CoroutineScope
) : MouseAdapter(), MouseMotionListener {
    private val DELAY = 300L

    // First and last directions from the player are differentiated so the task can be terminated when they have no direction in common,
    // in order to avoid an infinite loop
    private var firstDirectionsFromPlayer: List<Direction> = ArrayList()
    private var latestDirectionsFromPlayer: List<Direction> = ArrayList()
    private var entity: Entity? = null
    private val playerMovementTask = PeriodicTask({ movementTask() }, DELAY, scope)
    private var isMouseDragged = false
    private var isMouseClicked = false
    private var mouseClickQueue = LinkedList<MouseEvent>()
    private val player: Player
        get() = match.player!!

    var mouseCoords: Coordinates? = null
        private set

    /**
     * Retrieves and removes the next command in the mouseClickQueue.
     */
    private fun nextCommandInQueue() {
        mouseClickQueue.poll()

        if (!mouseClickQueue.isEmpty())
            onMouseClicked(mouseClickQueue.peekFirst())
    }

    private fun movementTask() {
        // Check if there is a player and the player is alive
        if (!player.logic.isAlive()) {
            return // If not, exit the task
        }

        onCooldown()

        entity?.let {
            // Perform mouse interactions on the entity
            it.logic.mouseClickedInteraction()
            onMovementPeriodicTaskEnd()
            nextCommandInQueue()
            return // Exit the task
        }

        // Calculate directions based on mouse coordinates
        latestDirectionsFromPlayer = mouseCoords?.fromCoordinatesToDirection(
                player.info.position
        ) ?: emptyList()

        // Press keys corresponding to calculated directions
        latestDirectionsFromPlayer.forEach {
            match.controllerManager?.onKeyPressed(it.toCommand())
        }

        // Check if first directions are not empty and all are in the latest directions
        if (firstDirectionsFromPlayer.isNotEmpty() && firstDirectionsFromPlayer.all { it in latestDirectionsFromPlayer }) {
            return // Exit the task
        }

        // Perform end-of-movement task operations
        onMovementPeriodicTaskEnd()
        nextCommandInQueue()
    }


    /**
     * Called when the mouse button is released.
     *
     * @param event The MouseEvent representing the release of the mouse button.
     */
    override fun mouseReleased(event: MouseEvent) {
        // If a mouse dragged interaction was not successful and the mouse is released, call mouseClicked
        // so that click-dependent interactions can be called.
        // This is necessary because mouseClicked is never called when mouseDragged is called, unlike mousePressed.
        // mouseClicked was still chosen because the two methods shouldn't be called at exactly the same time.
        if (!isMouseDragged) {
            return
        }

        mouseClicked(event)
        isMouseDragged = false
    }

    /**
     * Called when the mouse button is pressed.
     *
     * @param event The MouseEvent representing the press of the mouse button.
     */
    override fun mousePressed(event: MouseEvent) {
        if (isMouseClicked)
            return

        mouseCoords = Coordinates(event.x, event.y)
        entity = Coordinates.getEntityOnCoordinates(mouseCoords)
    }

    override fun mouseClicked(e: MouseEvent) {
        if (!mouseClickQueue.isEmpty()) {
            mouseClickQueue.add(e)
            return
        }

        mouseClickQueue.add(e)
        onMouseClicked(e)
    }

    private fun onMouseClicked(e: MouseEvent) {
        if (isMouseClicked)
            return

        isMouseClicked = true

        mouseCoords = Coordinates(e.x, e.y)
        firstDirectionsFromPlayer = mouseCoords!!.fromCoordinatesToDirection(
                player.info.position
        )

        entity = Coordinates.getEntityOnCoordinates(mouseCoords)
        startMouseTasks()
    }

    // Directions are refreshed and will be replaced in task
    private fun onCooldown() {
        val controllerManager = JBomb.match.controllerManager ?: return

        for (direction in firstDirectionsFromPlayer) {
            controllerManager.onKeyReleased(direction.toCommand())
        }
    }

    fun stopMovementTask() {
        mouseClickQueue = LinkedList()
        onMovementPeriodicTaskEnd()
    }

    private fun onMovementPeriodicTaskEnd() {
        if (!isMouseClicked)
            return

        playerMovementTask.stop()
        onCooldown()

        isMouseClicked = false
    }

    override fun mouseDragged(event: MouseEvent) {
        if (isMouseClicked)
            return

        isMouseDragged = true
        mouseCoords = Coordinates(event.x, event.y)

        entity?.logic?.mouseDraggedInteraction()
    }

    private fun startMouseTasks() {
        playerMovementTask.resume()
    }
}